### ISSUE TITLE
Improvement - Make SitRep effects on Tech research turn possible

### ISSUE THREAD

**Issue by agrrr3** [freeorion/freeorion#2355]
As well broadly discussed in [TurnTechResearched not working the first turn](https://www.freeorion.org/forum/viewtopic.php?f=9&t=11158&) it is not possible to send a custom sitrep in the same turn a technology gets researched (e.g. suitable planets for gas giant generators) due to the way the server does effect evaluation.

Gathering my current understanding I propose the following change:

~~- Introduce priority AFTER_RESEARCH_PRIORITY higher than anything we have currently~~
~~- Limit effect evaluation before processing research to effects with priority value lower than AFTER_RESEARCH_PRIORITY (adding priority min/max parameters to effect evaluation functions)~~
~~- Limit effect evaluation after processing research to effects with priority value higher than AFTER_RESEARCH_PRIORITY (adding priority min/max parameters to effect evaluation functions)~~
~~- Add extra tech initialization effects (currently i know only of sitreps for gas giant generators and  #2344 )~~

~~This is a low-risk approach moving effect evaluation in the direction of a priority based system.~~

Priorities and effects is a bit complicated so for the moment I rather suggest the approach implemented in #2370 : #2370 does not fiddle with priorities but removes the dubious first tech effect application so in effect tech does not behave differently for effect application between turns and is available consistently to FOCS coders.

I think the mentioned thread on the forum is ok for design discussion.

**Comment by agrrr3** [comment]
I started working on this and i am not sure the way is the right one. 
As one can see in Effect::EffectsGroup::Execute(..), the code has quite some history. 
```
Execute(const` TargetsCauses& targets_causes,
                    AccountingMap* accounting_map = nullptr,
                    bool only_meter_effects = false,
                    bool only_appearance_effects = false,
                    bool include_empire_meter_effects = false,
                    bool only_generate_sitrep_effects = false) const;
```
I guess the code was once concise and clear and then a lot necessities were discovered. So the effects' application is cut into different kind of effects which happens sometimes multiple times and leads to very confusing FOCS scripting .

I am not sure that ordering effects execution along priorities will clean this really up.
Maybe it does, but for the tech research case it might be possible to reduce complexity instead:
 just move the tech research to the end of the turn, so that it the only effect tech makes on the research turn is the sitrep informing one that the tech was successfully researched. None of our techs really needs initialization (like ships do).
The reduction of complexity stems from the fact that currently tech is researched in the middle of turn processing, leading to two different periods inside turn processing (effect application before tech is researched and effect application after tech is researched) instead of one (effect application is always after tech is researched).

@Dilvish-fo @adrianbroher @geoffthemedio i think you worked on effect application code most - what you think about the suggestion of moving tech research to the end of turn processing?

**Comment by agrrr3** [comment]
Note I have a [working implementation](https://github.com/agrrr3/freeorion/tree/TurnTechResearched_AFTER_RESEARCH_PRIORITY) of the AFTER_RESEARCH_PRIORITY idea.

**Comment by Dilvish-fo** [comment]
Hi, sorry I hadn't responded sooner, I've been a bit tied up.  The complication I don't see you acknowledging is that there are competing priorities of having the newly researched techs influencing the application of Effects (which is what it seems you are focused on) versus having the Effects contribute towards research, which I think the current placing had things.  (plus the current placing is consistent with when things are created.)

Simply saying that currently techs are researched in the middle of turn processing makes it sound muddled, which I don't think is the case, plus, because of how its organized and depending on how you look at it, any point could be considered 'in the middle'.  [As noted in our wiki](https://www.freeorion.org/index.php/FOCS_Scripting_Tutorial#Order_of_Events_During_the_Turn), the order of major turn processing events is (well sort of is, there is a note about updating for the changes to initial meter handling)

* orders are executed (invasions & colonizations done, fleet movement orders set, research & production queue changes set (but the progress on those queues is not yet handled)
* fleets_move and resupply takes place (includes refueling and also ship-part upgrades)
* empire visibility is updated & a pre-combat turn update is sent to players
* combat situations are determined and combat takes place
* empire visibility updated again
* loss of empire capitals checked
* all effects applied, including effect-specified sitrep generation, and including building new "current" * meter values from "initial" meter values
* empire visibility updated again
* blockades, supply & resources updated
* things made, __techs researched__
* scripted events processed
* meter effects applied again (rebuilds new "current" meter values from "initial" meter values, not directly adding another dollop to the previous iteration of current meter values). Note: that this does not currently include an updated round of effect-specified sitrep generation and so discrepancies can arise from that.
* population growth
* back-propagation of meters ("initial" values overwritten by "current" values)
* empire visibility updated again
* turn number advances
* empire visibility updated again (this time it counts for the new turn)
* post-combat turn update is sent to players
* players enter orders, cycle repeats

I see you have a PR up now, and this discussion could go either place, but since this aspect leans towards the planning/design side of things I think it fits a bit better here.  Could you clarify just where it is in the above order of processing it is that you are proposing to move the researching of techs to?

**Comment by agrrr3** [comment]
Hi @Dilvish-fo , thanks for the reply

the underlying problem as i would phrase it is that tech effects work differently in the turn the tech gets researched from all other turns and that this leads to complications for FOCS scripters. 

> The complication I don't see you acknowledging is that there are competing priorities of having the newly researched techs influencing the application of Effects (which is what it seems you are focused on) versus having the Effects contribute towards research, which I think the current placing had things. (plus the current placing is consistent with when things are created.)

I am not sure I understand what you write. Do you mean i do not address that effects influence the research points used for researching tech?
In #2370 i suggest to split these issues. Tech research is calculated as is, but the tech is actually added later to the universe, so tech effects will not be different between the first researched turn and the other turns.

> 
> Simply saying that currently techs are researched in the middle of turn processing makes it sound muddled, which I don't think is the case, plus, because of how its organized and depending on how you look at it, any point could be considered 'in the middle'. 

If it does not have an effect, i would not consider it as 'in the middle'. Would you?

What i suggest in #2370 would change the turn processing as follows:
> 
>     * orders are executed (invasions & colonizations done, fleet movement orders set, research & production queue changes set (but the progress on those queues is not yet handled)
> 
>     * fleets_move and resupply takes place (includes refueling and also ship-part upgrades)
> 
>     * empire visibility is updated & a pre-combat turn update is sent to players
> 
>     * combat situations are determined and combat takes place
> 
>     * empire visibility updated again
> 
>     * loss of empire capitals checked
> 
>     * all effects applied, including effect-specified sitrep generation, and including building new "current" * meter values from "initial" meter values
> 
>     * empire visibility updated again
> 
>     * blockades, supply & resources updated
> 
>     * things made, **techs researched**

Instead of adding researched techs to the empires, remember which techs are researched.

> 
>     * scripted events processed
> 
>     * meter effects applied again (rebuilds new "current" meter values from "initial" meter values, not directly adding another dollop to the previous iteration of current meter values). Note: that this does not currently include an updated round of effect-specified sitrep generation and so discrepancies can arise from that.

So no meter effects of the innovations, as they are not yet added to the empires.

> 
>     * population growth
> 
>     * back-propagation of meters ("initial" values overwritten by "current" values)

About here i add the techs to the empires.

> 
>     * empire visibility updated again
> 
>     * turn number advances
> 
>     * empire visibility updated again (this time it counts for the new turn)
> 
>     * post-combat turn update is sent to players
> 
>     * players enter orders, cycle repeats
> 
> 

I am not sure this is the right place to discuss the approach of #2370 , as i originally suggested here another one which introduces priority-based filtering to make fine-grained effect application possible.

**Comment by agrrr3** [comment]
@Dilvish-fo bump?

**Comment by agrrr3** [comment]
@Dilvish-fo ?

**Comment by Vezzra** [comment]
Hm, looks like @Dilvish-fo is MIA again... can anyone else help out with this...?

**Comment by agrrr3** [comment]
@Vezzra @geoffthemedio As @Dilvish-fo is MIA this stalled and I am waiting since six months for a reply if the approach in #2370 is ok.

**Comment by Vezzra** [comment]
Unfortunately I can't really give a competent opinion here - @geoffthemedio? Seems like you'll have to make this call, I don't know who else could...

**Comment by geoffthemedio** [comment]
@agrrr3 Resolved now?

**Comment by agrrr3** [comment]
Resolved with #2370 merged
